### PR TITLE
Fix unsoundness in read_to_end

### PR DESCRIPTION
**Pull request by Darksonn**
This is a companion of #3426 that contains just the unsoundness fix so we can release it in a patch release.

### CODE REVIEW

**Review by carllerche — APPROVED** [review]
Thanks :+1:

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,8 +1,8 @@
 on:
   push:
-    branches: ["master"]
+    branches: ["tokio-1.0.x"]
```
Comment (carllerche) [inline .github/workflows/ci.yml:3]: Keep master as we will eventually merge into master. Also, we should be able to do a glob: "tokio-*.x".

Comment (Darksonn) [inline .github/workflows/ci.yml:3]:
```suggestion
    branches: ["master", "tokio-*.x"]
```

Comment (carllerche) [reply .github/workflows/ci.yml:3]: Example: https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#example-including-branches-and-tags

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -1,8 +1,8 @@
...
   pull_request:
-    branches: ["master"]
+    branches: ["tokio-1.0.x"]
```
Comment (Darksonn) [inline .github/workflows/ci.yml:5]:
```suggestion
    branches: ["master", "tokio-*.x"]
```

**Pull request merged**